xskillscore/tests/test_accessor_deterministic.py

```diff
@@ -0,0 +1,173 @@
+    rmse,
+    smape,
+    spearman_r,
+    spearman_r_p_value,
+    spearman_r_eff_p_value,
+    effective_sample_size,
+    r2,
+)
+
+
+correlation_metrics = [
+    pearson_r,
+    r2,
+    pearson_r_p_value,
+    spearman_r,
+    spearman_r_p_value,
+    effective_sample_size,
+    pearson_r_eff_p_value,
+    spearman_r_eff_p_value,
+]
+
+temporal_only_metrics = [
+    pearson_r_eff_p_value,
+    spearman_r_eff_p_value,
+    effective_sample_size,
+]
+
+distance_metrics = [
+    mse,
+    rmse,
+    mae,
+    median_absolute_error,
+    mape,
+    smape,
+]
+
+AXES = ("time", "lat", "lon", ["lat", "lon"], ["time", "lat", "lon"])
+
+
+@pytest.fixture
+def a():
+    times = pd.date_range("1/1/2000", "1/3/2000", freq="D")
+    lats = np.arange(4)
+    lons = np.arange(5)
+    data = np.random.rand(len(times), len(lats), len(lons))
+    return xr.DataArray(data, coords=[times, lats, lons], dims=["time", "lat", "lon"])
+
+
+@pytest.fixture
+def b(a):
+    b = a.copy()
+    b.values = np.random.rand(a.shape[0], a.shape[1], a.shape[2])
+    return b
+
+
+@pytest.fixture
+def weights(a):
+    """Weighting array by cosine of the latitude."""
+    a_weighted = a.copy()
+    cos = np.abs(np.cos(a.lat))
+    data = np.tile(cos, (a.shape[0], a.shape[2], 1)).reshape(
+        a.shape[0], a.shape[1], a.shape[2]
+    )
+    a_weighted.values = data
+    return a_weighted
+
+
+def _ds(a, b, skipna_bool, dask_bool):
+    ds = xr.Dataset()
+    ds["a"] = a
+    ds["b"] = b
+    if skipna_bool is True:
+        ds["b"] = b.where(b < 0.5)
+    if dask_bool is True:
+        ds["a"] = a.chunk()
+        ds["b"] = b.chunk()
+    return ds
+
+
+def adjust_weights(dim, weight_bool, weights):
+    """
+    Adjust the weights test data to only span the core dimension
+    that the function is being applied over.
+    """
+    if weight_bool:
+        drop_dims = [i for i in weights.dims if i not in dim]
+        drop_dims = {k: 0 for k in drop_dims}
+        return weights.isel(drop_dims)
+    else:
+        return None
+
+
+def get_func(ds, metric):
+    if metric == pearson_r:
+        return ds.xs.pearson_r
+    if metric == r2:
+        return ds.xs.r2
+    if metric == pearson_r_p_value:
+        return ds.xs.pearson_r_p_value
+    if metric == spearman_r:
+        return ds.xs.spearman_r
+    if metric == spearman_r_p_value:
+        return ds.xs.spearman_r_p_value
+    if metric == effective_sample_size:
+        return ds.xs.effective_sample_size
+    if metric == pearson_r_eff_p_value:
+        return ds.xs.pearson_r_eff_p_value
+    if metric == spearman_r_eff_p_value:
+        return ds.xs.spearman_r_eff_p_value
+    if metric == mse:
+        return ds.xs.mse
+    if metric == rmse:
+        return ds.xs.rmse
+    if metric == mae:
+        return ds.xs.mae
+    if metric == median_absolute_error:
+        return ds.xs.median_absolute_error
+    if metric == mape:
+        return ds.xs.mape
+    if metric == smape:
+        return ds.xs.smape

```

Comment:
Anyone know how to avoid this?

Reply:
Something like this:

```python
func_dict = {pearson_r: ds.xs.pearson_r, r2: ds.xs.r2, ...}
def get_func(ds, metric):
    try:
        return func_dict[metric]
    except KeyError:
        raise ValueError(f'{metric} not in {func_dict.keys()})
```

Reply:
Thanks.
Cleaner than the multiple if statements.
Was trying ds.xs.METRIC at first but your can't declare it like that.

Reply:
consider using `getattr(xs,metric_str)`

Reply:
> consider using `getattr(xs,metric_str)`

Thanks. Looking at https://docs.python.org/3/library/functions.html#getattr